src/AutoMerge/util.jl

```diff
@@ -84,6 +86,192 @@ function parse_registry_pkg_info(registry_path, pkg, version=nothing)
+                end
+            end
+        catch e
+            @warn e
+        end
+        return nothing
+    else
+        # Only commits that touched `subdir` (much smaller set)
+        commits_s = try
+            readchomp(Cmd(`git -c gc.auto=0 log --all --format=%H -- $subdir`;
+                          dir=clone_dir, env=Dict("GIT_NO_REPLACE_OBJECTS"=>"1")))
+        catch e
+            @warn e
+            return nothing
+        end
+        isempty(commits_s) && return nothing
+        commits = split(commits_s, '\n'; keepempty=false)
+
+        # Check subdir tree per candidate commit (fast enough in practice)
+        for c in commits
+            isempty(c) && continue
+            t = try
+                readchomp(Cmd(`git rev-parse $c:$subdir`;
+                              dir=clone_dir, env=Dict("GIT_NO_REPLACE_OBJECTS"=>"1")))
+            catch
+                continue  # subdir may not exist at this commit
+            end
+            if t == full_tree
+                return c
+            end
+        end
+        return nothing
+    end
+end
+
+"""
+    is_github_repo(repo_url::String) -> Bool
+
+Checks if a repository URL is a GitHub repository.
+"""
+function is_github_repo(repo_url::String)
+    return occursin(r"github\.com[:/]", repo_url)
+end
+
+"""
+    extract_github_owner_repo(repo_url::String) -> Union{Tuple{String, String}, Nothing}
+
+Extracts the owner and repository name from a GitHub URL.
+Returns a tuple (owner, repo) or `nothing` if the URL is not a valid GitHub URL.
+"""
+function extract_github_owner_repo(repo_url::String)
+    # Handle both HTTPS and SSH GitHub URLs
+    # HTTPS: https://github.com/owner/repo.git
+    # SSH: git@github.com:owner/repo.git
+    m = match(r"github\.com[:/]([^/]+)/([^/]+?)(?:\.git)?/?$", repo_url)
+    return m === nothing ? nothing : (m.captures[1], m.captures[2])
+end
+
+"""
+    generate_github_diff_url(repo_url::String, previous_commit_sha::String, current_commit_sha::String) -> Union{String, Nothing}
+
+Generates a GitHub diff URL comparing two commits.
+Returns the URL string, or `nothing` if the repository is not on GitHub.
+"""
+function generate_github_diff_url(repo_url::String, previous_commit_sha::String, current_commit_sha::String)
+    if !is_github_repo(repo_url)
+        return nothing
+    end
+
+    owner_repo = extract_github_owner_repo(repo_url)
+    if owner_repo === nothing
+        return nothing
+    end
+
+    owner, repo = owner_repo
+    return "https://github.com/$owner/$repo/compare/$previous_commit_sha...$current_commit_sha"
+end
+
+"""
+    get_version_diff_info(data) -> Union{NamedTuple, Nothing}
+
+Gets diff information for a new version registration. Returns a NamedTuple with fields:
+- `diff_url`: GitHub diff URL
+- `previous_version`: Previous version number
+- `current_version`: Current version number
+
+Returns `nothing` if no previous version exists or if the repository is not on GitHub.
```

Comment:
this potentially allows attackers to avoid the diff surfacing by doing something bad to the repo (rewriting history to remove the old tree for example). however, if the diff feature works well and is reliable (and we add gitlab support), we could later move it to a required check (i.e. we must be able to successfully generate a diff link) for new versions. So IMO this is a first step.